### ISSUE TITLE
Unnecessary transitive dependencies now limiting our ability to move independently of unrelated projects

### ISSUE THREAD

**Issue by ironcladlou** [openshift/hypershift#55]
Unnecessary transitive dependencies have crept in and are now limiting our ability to move independently of unrelated projects.

HyperShift depends on several upstream libraries using Go modules, including:

* k8s.io/client-go
* sigs.k8s.io/controller-runtime
* k8s.io/api

HyperShift also depends on the following Go modules for the purpose of having Go types representing the third-party CRD APIs we leverage:

* sigs.k8s.io/cluster-api
* sigs.k8s.io/cluster-api-provider-aws

It's important to note the only reason we declare these API dependencies  is to gain use of the Go types in our codebase. **This is convenient but is in no way a technical constraint.** For example, if we really wanted to, we could use unstructured types to achieve the same effect.

The CRD API modules we depend on are not structured as Go submodules; therefore, when HyperShift declares a Go dependency on these API modules, HyperShift inherits those projects' own internal dependencies transitively. **As a result, we currently have irreconcilable conflicting versions of the machinery libraries listed above.**

The net effect is our application has imposed upon it serious limitations by unrelated third-party applications. For example, the version of the Kubernetes API client or controller-runtime we use in HyperShift should not be in any way dictated by the fact that an unrelated third-party application (e.g. cluster-api-provider-aws) also uses a conflicting version controller-runtime.

**This coupling of application dependencies is unacceptable and must be eliminated.** The upstream projects which seek to provide publicly consumable API types as Go modules need to make changes to separate these very different concerns:

1. Providing public API types
2. Providing an application that is a consumer of the API types

### Options

We have options for regaining control of our dependency graph. Here are some I thought of:

1. Use unstructured types instead of the upstream Go types.
1. Convince the upstream projects to extract their APIs into new Go modules which themselves declare no unreasonable external dependencies.
    - Example: Separate API repositories
    - Example: Nested Go modules in the same repository
1. Copy and keep in sync the Go API types we want from their upstream sources.
    - Some of the existing upstream API types are also doing anti-consumer things like declaring dependencies on k8s machinery and controller-runtime interfaces, so a degree of post-processing is inevitable right now with this approach.
1. Generate our own internal versions of the Go API types from the OpenAPI schemas provided by the upstream projects.
    - https://github.com/OpenAPITools/openapi-generator
    - https://github.com/deepmap/oapi-codegen
      - https://github.com/deepmap/oapi-codegen#whats-missing-or-incomplete
    - Others?

### Beyond

After the immediate problems are resolved, we must remain dilligent with our dependencies going forward, taking care not to stumble back into a similar situation. Unfortunately, Go modules and the communities that use them make it very easy to inadvertantly end up in a web of unintended transitive dependencies. One mitigation is to simply agree on a principle that _all changes that affect the dependency graph_ should be carefully scrutinized and viewed through a skeptical lens (i.e. "do we really understand the implications of accepting this dependency?")

### Postscript: sigs.k8s.io/cluster-api/util/patch

The use of the `sigs.k8s.io/cluster-api/util/patch` package has found its way into the codebase, and for understandable reasons (it seems to be useful in some contexts, DRY, etc.) However, regardless of how useful it is, the fact remains it is published as part of the `sigs.k8s.io/cluster-api` module and so brings along all the problems described above with the API types. If this utility is truly desirable for controller authors generally, the code should be packaged and published with public library consumers in mind: in a separate Go submodule with a minimal/narrow set of dependencies, in a separate repo, or incorporated into an existing upstream project focused on enabling controller authors (e.g. controller-runtime).

Until then, **the `sigs.k8s.io/cluster-api/util/patch` dependency must be eliminated one way or another**, e.g. copying the code, replacing it with controller-runtime equivalents, persuading upstream to change how it's published, etc.

**Comment by ironcladlou** [comment]
cc @csrwng @sjenning @enxebre @derekwaynecarr @elmiko

**Comment by ironcladlou** [comment]
If you want to see where this has led us, try to update the following dependencies on `main`:

```
k8s.io/api v0.20.2
k8s.io/apiextensions-apiserver v0.20.2
k8s.io/apimachinery v0.20.2
k8s.io/client-go v0.20.2
sigs.k8s.io/controller-runtime v0.8.2
```

And observe the fallout through trying to compile (see `go mod graph`)

**Comment by ironcladlou** [comment]
Regarding this approach:

>Generate our own internal versions of the Go API types from the OpenAPI schemas provided by the upstream projects.
https://github.com/OpenAPITools/openapi-generator
https://github.com/deepmap/oapi-codegen
https://github.com/deepmap/oapi-codegen#whats-missing-or-incomplete

I've played with OpenAPITools/openapi-generator some, and the problems I've encountered are serious:

* Getting the v3 openapi schema for any types at all from the cluster isn't implemented (only v2 is accessible)
* The v2 types from the server, assuming that's even useful, (i.e. `oc get --raw="/openapi/v2"`) implies we install the CRD into a cluster so we can ask the server for the generated schema, and the result needs further processed to extract just the types we want
* I manually constructed a valid v2 schema for CAPI Cluster type as an experiment to run through the generator, and the output is pretty crazy and not useful as-is

So far it seems like a dead end compared to just copying in the Go types we want.

**Comment by elmiko** [comment]
thanks for the detailed explanation @ironcladlou , i'm going to bring up a discussion of modules and openapi generated code in the capi meeting this week. is there any extra detail you could give about the manually constructed schema you created, what were the sticking points? anything you would like us to highlight with upstream?

**Comment by ironcladlou** [comment]
@elmiko 
>thanks for the detailed explanation @ironcladlou , i'm going to bring up a discussion of modules and openapi generated code in the capi meeting this week. is there any extra detail you could give about the manually constructed schema you created, what were the sticking points? anything you would like us to highlight with upstream?

The problems I had with the OpenAPI->Go approach were more to do with k8s itself. There's no tooling I can find to go from a serialized CRD to OpenAPIv3 schema suitable for processing by any tools I've found so far. And the apiserver will only present me with a complete OpenAPIv2 spec for every type in the cluster. It's the latter I was able to pick apart and feed into tools. The output I got from OpenAPITools/openapi-generator this way was so enormous and complex I didn't really bother trying to figure out whether it could work or be adapted. Maybe with more time and experimentation it could be made to work through some complicated pipeline.

I'm not sure what would be the "ideal" yet from an API publisher perspective. Maybe it means Go submodules (nested or in a separate repo), publishing the raw OpenAPI spec JSON for generator pipelines, etc. It would be useful if somebody had an example of this done "correctly". The https://github.com/openshift/api project is much closer to what I have in mind as sane. Here's its complete `go.mod` declaration:

```
require (
	github.com/gogo/protobuf v1.3.1
	github.com/openshift/build-machinery-go v0.0.0-20200917070002-f171684f77ab
	github.com/spf13/pflag v1.0.5
	golang.org/x/tools v0.0.0-20200616133436-c1934b75d054
	k8s.io/api v0.20.0
	k8s.io/apimachinery v0.20.0
	k8s.io/code-generator v0.20.0
	k8s.io/klog/v2 v2.4.0
)
```

As an API consumer this is far closer to what I expect from an API package in terms of transitive conflict potential. I'd go even further and say `klog` and `pflag` and `openshift/build-machinery-go` should be removed as well.
